### PR TITLE
add comment-no-empty rule

### LINKED ISSUE #1677: comment-no-empty
Disallow empty comments.

Invalid:
```css
/* */
div {
  /* */
  color: red;
  /* */
}
/* */
```

Valid:
```css
/* Some comment */
div {
  /* Some comment */
  color: red;
 /* display: block; */
 /* Some comment */
}
/* Some comment */
```

### PR DESCRIPTION
**Pull request by sendilkumarn**
fixes #1677 

is it really this `simple` or am i `missing` something here?

### CODE REVIEW

**Comment by evilebottnawi** [comment]
@jeddy3 That do you think about comment contain only `\n`, `\r\n` and `\t` (multiples also)?

**Comment by evilebottnawi** [comment]
@sendilkumarn https://github.com/stylelint/stylelint/blob/master/src/rules/block-no-empty/__tests__/index.js#L38 `\n` should accepted and other spaces characters too

**Comment by sendilkumarn** [comment]
Accepted as valid is it ?

**Comment by evilebottnawi** [comment]
@sendilkumarn comment contain `\n` and other spaces characters considered `empty`.

**Comment by sendilkumarn** [comment]
yeah `exactly` they are considered as empty.

**Comment by jeddy3** [comment]
@sendilkumarn Can you add the same whitespace tests from https://github.com/stylelint/stylelint/pull/1684 into here please.

You'll need to rebase too.

File: src/rules/comment-no-empty/README.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# comment-no-empty
+
+Disallow empty comments.
+
+```css
+ a { /* */ display: block; }
```
Comment (jeddy3) [inline src/rules/comment-no-empty/README.md]: There's a lot of distraction in the docs and tests.  Lets simplify them down e.g.

```css
    /* */
/** ↑
 * Comments like this */
```

File: src/rules/comment-no-empty/README.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+```
+
+## Options
+
+### `true`
+
+The following patterns are considered warnings:
+
+```css
+a {
```
Comment (jeddy3) [inline src/rules/comment-no-empty/README.md]:
```css
/**/
```

File: src/rules/comment-no-empty/README.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  /* */
+  display: block;
+}
+```
+
+```css
```
Comment (jeddy3) [inline src/rules/comment-no-empty/README.md]:
```css
/* */
```

File: src/rules/comment-no-empty/README.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+/* */ a { display: block; }
+```
+
+```css
```
Comment (jeddy3) [inline src/rules/comment-no-empty/README.md]:
```css
/*

*/
```

File: src/rules/comment-no-empty/README.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+@media print { /* */ a {} }
+```
+
+The following patterns are *not* considered warnings:
+
+```css
```
Comment (jeddy3) [inline src/rules/comment-no-empty/README.md]:
```css
/* Some Comment */
```

File: src/rules/comment-no-empty/README.md [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+a {  /* Some Comment */ color: pink; }
+```
+
+```css
```
Comment (jeddy3) [inline src/rules/comment-no-empty/README.md]:
```css
/*
 * Some comment
**/
```

File: src/rules/comment-no-empty/__tests__/index.js [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+import { testRule } from "../../../testUtils"
+
+const rule = rules[ruleName]
+
+testRule(rule, {
+  ruleName,
+  config: [true],
+  skipBasicChecks: true,
+
+  accept: [ {
```
Comment (jeddy3) [inline src/rules/comment-no-empty/__tests__/index.js]: The focus of these tests should be comments and what's considered empty. It's not clear what's being tested as there's two much non-comment code

File: src/rules/comment-no-empty/__tests__/index.js [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+} from ".."
+import rules from "../../../rules"
+import { testRule } from "../../../testUtils"
+
+const rule = rules[ruleName]
+
+testRule(rule, {
+  ruleName,
+  config: [true],
+  skipBasicChecks: true,
```
Comment (jeddy3) [inline src/rules/comment-no-empty/__tests__/index.js]: Remove `skipBasicChecks: true,` as the basicChecks should be used here.

File: src/rules/comment-no-empty/index.js [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+      // Check if the comment text is empty
+      if (comment.text && comment.text.length !== 0) { return }
+
+      report({
+        message: messages.rejected,
+        node: comment,
+        result,
+        ruleName,
+      })
+
```
Comment (jeddy3) [inline src/rules/comment-no-empty/index.js]: Remove empty lines

File: src/rules/comment-no-empty/index.js [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+  rejected: "Unexpected empty comment",
+})
+
+export default function (actual) {
+  return (root, result) => {
+    const validOptions = validateOptions(result, ruleName, { actual })
+    if (!validOptions) { return }
+
+    root.walkComments(comment => {
+      // Check if the comment text is empty
```
Comment (jeddy3) [inline src/rules/comment-no-empty/index.js]: You'll want to [ignore inline SCSS comments](https://github.com/stylelint/stylelint/blob/master/src/rules/comment-empty-line-before/index.js#L57) here.

**Comment by sendilkumarn** [comment]
@jeddy3 `postcss` doesn't seem to parse the following comments correctly.

```css
a, b /**/ { } 
```

so `removed` that test case alone.

**Comment by evilebottnawi** [comment]
@sendilkumarn yep, i think in postcss v 5.* we can write utils which collects this comments, if we have time on this :smile:

**Comment by jeddy3** [comment]
> yep, i think in postcss v 5.* we can write utils which collects this comments

@sendilkumarn What @evilebottnawi is referring to that these comments are available in the `raws` property of the AST.

```css
selector, /**/
list
/**/ {
  prop: value, /**/ list;
}
```

The *selector list* comment is available in `rule.raws.selector.raw` and the *value list* one in `decl.raws.value.raw`. See https://astexplorer.net/#/98aeoVDkEw.

@stylelint/core We need to decide whether:

1. We're going to write these utils now. 
2. Or just document the limitation in the README and wait for PostCSS 6.0 to address the issue.

(This decision is also for https://github.com/stylelint/stylelint/issues/1628).

I'm leaning towards the latter as if waiting avoids work at our end, then I'm all for that.

**Comment by davidtheclark** [comment]
What about using `styleSearch` to find comments instead of the AST?

Could you use `styleSearch` to find `/*`, then `balanced-match` to get the text between that and the closing `*/`, then check whether that content is only whitespace?

**Comment by jeddy3** [comment]
> Could you use `styleSearch` to find `/*`, then `balanced-match` to get the text between that and the closing `*/`, then check whether that content is only whitespace?

That would definitely work for this rule, as it's new. Using a similar approach for #1628 would be a breaking change as `// stylelint-disable` commands are currently accepted. So, we'd probably end up having to write the `raws` utils for that anyway.

Or, we use `style-search` here (it it'll make this rule feature-complete) and then just document the limitation in #1628 (as it can be easily be worked around by using `stylelint-disable` with `stylelint-enable`, rather than `stylelint-disable-line`)?

**Comment by davidtheclark** [comment]
`//` comments should also be detected by `styleSearch`, and you can tell where they end by looking for the next newline.

**Comment by jeddy3** [comment]
>`//` comments should also be detected by `styleSearch`, and you can tell where they end by looking for the next newline.

That's a good point.

One more data point in our discussion is that the existing `comment-*` rules also have this limitation e.g.

```js
{
  "comment-empty-line-before": "always",
  "comment-whitespace-inside": "always",
  "comment-word-blacklist" : ["TODO"]
}
```

```css
selector, 
/*TODO*/
list {
  prop: value, 
    /*TODO*/ 
    list;
}
```

Will currently produce no warnings.

Is it better to?:

1. Have *a consistent* limitation across all the `comment` rules that are fixed in one go in the future (either my swapping to `style-search` or waiting for PostCSS to fix the AST in `6.0`).
2. Or to make this rule (and future `comment` rules) complete and fix the `stylelint-disable` commands (as that's not breaking) now by swapping to `style-search`, and then fix the existing `comment` rules in `future-major` (again by swapping to `style-search`).

No clear right answer. I'm still leaning towards documenting a consistent limitation and waiting on a upstream fix. If only because it's consistent and less effort, but I'm easy either way if someone wants to put the effort in :)

**Comment by sendilkumarn** [comment]
I would second @jeddy3 

not that it is about re-work. If postcss will be having this `fix` in future. We will `wait`.

Why we need to fix something to new part only. It will result in inconsistent.

We will look at how to fix at `postcss` side

**Comment by davidtheclark** [comment]
Ok. Too bad we didn't catch it earlier.

**Comment by sendilkumarn** [comment]
Yeah we should have captured it earlier.

File: src/rules/comment-no-empty/README.md (hunk 1 above)
Comment (jeddy3) [inline src/rules/comment-no-empty/README.md]: Can you change the proto-example please.

File: src/rules/comment-no-empty/README.md (hunk 2 above)
Comment (jeddy3) [inline src/rules/comment-no-empty/README.md]: And this one.

File: src/rules/comment-no-empty/README.md (hunk 5 above)
Comment (jeddy3) [inline src/rules/comment-no-empty/README.md]: And this one.

**Comment by jeddy3** [comment]
> Yeah we should have captured it earlier.

Ha, I think we can be a bit more forgiving of ourselves. Not realising PostCSS's `walkComments` doesn't capture selector and value list comments is kind of forgiveable. I mean, it's only just popped up in an issue and it's a year down the line :) I agree it's a shame we didn't catch it though.

@sendilkumarn If you can make those three documentation changes I've requested, I think this is ready to go in?

I'll document the limitation across the board in another PR.

**Comment by sendilkumarn** [comment]
What is that 'proto-example'

I did not get that sorry..

**Comment by jeddy3** [comment]
The first example e.g.

---

# comment-no-empty

Disallow empty comments.

```css
    /* */
/** ↑
 * Comments like this */
```

---

**Comment by sendilkumarn** [comment]
@jeddy3 @davidtheclark let me know if you have any more comments on this

**Comment by jeddy3** [comment]
LGTM

File: src/rules/comment-no-empty/__tests__/index.js (hunk 8 above)
Comment (davidtheclark) [inline src/rules/comment-no-empty/__tests__/index.js]: @sendilkumarn Did you have a reason *not* to remove this?

**Comment by davidtheclark** [comment]
@sendilkumarn Did you miss @jeddy3's comment about `skipBasicChecks`?

**Comment by davidtheclark** [comment]
Great! Thanks.

**Comment by jeddy3** [comment]
@sendilkumarn Thanks for this!

**Pull request merged**